rust/lance-index/src/vector/graph/memory.rs

```diff
@@ -80,10 +80,36 @@ struct InMemoryDistanceCal {
     metric_type: MetricType,
 }
 
-impl DistCalculator for InMemoryDistanceCal {
+impl DistCalculator<'_> for InMemoryDistanceCal {
     #[inline]
     fn distance(&self, id: u32) -> f32 {
         let vector = self.vectors.row_ref(id as usize).unwrap();
         self.metric_type.func()(self.query.as_primitive::<Float32Type>().values(), vector)
     }
+    fn prefetch(&self, id: u32) {
+        // TODO use rust intrinsics instead of x86 intrinsics
+        // TODO finish this
+        unsafe {
+            let vector = self.vectors.row_ref(id as usize).unwrap();
+            let ptr = vector.as_ptr() as *const i8;
+            let end_ptr = vector.as_ptr().add(vector.len()) as *const i8;
+
+            let mut current_ptr = ptr;
+            while current_ptr < end_ptr {
+                const CACHE_LINE_SIZE: usize = 64;
+                #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+                {
+                    use core::arch::x86_64::_mm_prefetch;
+                    const _MM_HINT_T0: i32 = 1;
```

Comment:
let's just import the rust const here: https://doc.rust-lang.org/beta/core/arch/x86_64/constant._MM_HINT_T0.html

rust uses different numbers compare to c/c++

Reply:
I'll add for info to Yang here: We tested this. The cache level to fetch to doesn't end up making a noticeable difference (probably because an l1 cache hit vs an l2 cache hit is unnoticeable compared to the computation time). But yeah we should have `use core::arch::x86_64::_MM_HINT_T0` instead of a manual const declaration (because _MM_HINT_T0 is not 1, it's 3, and they might also change that in the future).